### DISCUSSION TITLE
Hiding a pipeline resource variable from certain shaders

### DISCUSSION THREAD

**Discussion by RobertMtx** [DiligentGraphics/DiligentEngine#260] [Q&A]
In my engine, every single shader links to the same global pipeline resource signature, which is pretty convenient. However, my shadowmap rendering vertex shaders are complaining with the following output when I render to my shadowmap depth texture:

> Diligent Engine: ERROR: Texture 'Shadow Map' must be in RESOURCE_STATE_SHADER_RESOURCE state. Actual state: DEPTH_WRITE. Call IDeviceContext::TransitionShaderResources(), use RESOURCE_STATE_TRANSITION_MODE_TRANSITION when calling IDeviceContext::CommitShaderResources() or explicitly transition the texture state with IDeviceContext::TransitionResourceStates().

For some reason, this message only shows up once during the first shadowmap render, then only shows up again if several frames are skipped before another one is rendered. I'm assuming this is Diligent Engine trying to avoid flooding the output with messages, but haven't been able to confirm that. If not, it may indicate that I'm doing something specific during that frame that is causing an issue.

Another assumption I have is that this message shows up because the shadowmap is a texture resource in the shadowmap rendering vertex shader's resource signature (because all shaders share that signature). Even though the shadow rendering shader is actually writing to this texture as a render target, the resource signature apparently still wants to verify it is in shader resource mode, even with the shader not actually touching the resource.

My question is - is there a way to resolve this? Can I hide the shadowmap texture resource from these specific vertex shaders? Or will I be forced to create another global resource signature for shadow rendering shaders, which is identical in every way except missing that texture? Is there some other way?

EDIT: I believe this error showing up (or not) has something to do with when `CommitShaderResources()` is called. If I typically call `CommitShaderResources()` at the very beginning of my frames, I only get the error once (on the first shadow render, then again any time I render shadows again, after stopping for a while). But if I call `CommitShaderResources()` (still once per frame) right before the first call to `Context.DrawIndexed()`, I will get the error every single frame. Any idea why?

EDIT 2: If I call `CommitShaderResources()` (still only once per frame) right before I set my shadowmap depth buffer render target, I *never* get the error.  But if I move it to the line right below the render target assignment (so it happens after), I get the error every frame. These results have left me pretty confused.

EDIT 3: Last one, I promise. I just wanted to mention that the reason I was getting a single error message when I stopped rendering shadowmaps is because my environment was being unloaded/reloaded, and its environment textures were being set, forcing `CommitShaderResources()` to be called that frame **after** the shadow render target was set. So yeah, this seems to be root of the error. It just depends on if my commit is called before or after I set shadowmap render targets. But I don't understand the reason for it. Why is it happy as long as I commit before I set the shadowmap as a render target? Is it because its reading the state of the resource *at the time it was committed*? That might make sense, but it would mean I'm still doing it wrong even when I don't get the error, but Diligent can't detect it.

Appreciate any advice!

**Comment by TheMostDiligent** [comment]
When you call `CommitShaderResources()`, the engine transitions all resources in the SRB to shader resource state. If you commit the SRB with shadow map resource before you render the shadow map, then the shadow map will be transitioned to the DEPTH_WRITE state when you render to it, so that when you later use it as resource, there will be the mismatch.
In this case you will need to manually transition the shadow map to the SHADER_RESOURCE state after you render it.

**Comment by RobertMtx** [comment]
Well, the vertex shader that is active at the time of the error actually needs the texture to be in a render target state. The texture resource is present as a global variable and is part of the resource signature being used by the shadow rendering vertex shaders, but it isn't touched by those shadow rendering vertex shaders.

> When you call CommitShaderResources(), the engine transitions all resources in the SRB to shader resource state.

Is that true even if I use Diligent::RESOURCE_STATE_TRANSITION_MODE_VERIFY only? I'm only using verify in my call. I manually transition my texture between render target and shader resource. Something like this:

1. Transition shadow texture to render target state
2. Set shadow texture as render target depth buffer
3. Activate shadow rendering vertex shaders / no pixel shaders (this is where I get the error)
4. Set shadow texture as shader resource
5. Render standard graphics that include pixel shaders, which make use of shadow texture as a shader resource for the first time

**Reply by TheMostDiligent** [reply]
> Is that true even if I use Diligent::RESOURCE_STATE_TRANSITION_MODE_VERIFY only?

No, if you use the `VERIFY` mode, it will obviously verify the states only.
When you call `CommitShaderResources(VERIFY)`, the engine verifies that all resources in the SRB are in `SHADER_RESOURCE` state. It can't know which resources will actually be needed by the PSO, so it verifies all. If some of the resources are not in `SHADER_RESOURCE` state, it will print a warning, which may be a false positive. In this case you need to disable verification and use the `NONE` mode.

> Activate shadow rendering vertex shaders / no pixel shaders (this is where I get the error)

Do you mean when you call the `SetPipelineState`? `SetPipelineState` should only check the resources that are actually used by the PSO.

**Reply by RobertMtx** [reply]
> Do you mean when you call the SetPipelineState? SetPipelineState should only check the resources that are actually used by the PSO.

No, I check for the need to call `CommitShaderResources()` right before any poly draw, and the error is triggered while execution is inside of `CommitShaderResources()`. But the poly draw that's about to happen when the error is triggered is a shadow rendering shader that does not use the texture in question as a shader resource. Is there anything I can do to resolve the error?

> SetPipelineState should only check the resources that are actually used by the PSO.

The texture its referring to is set as a depth-write render target when the error is triggered. While the texture is visible as a global variable, the shader does not use it. Would it help to wrap the variable in an `#ifdef` that is turned off for shadow map vertex shaders?

**Reply by TheMostDiligent** [reply]
> Would it help to wrap the variable in an #ifdef that is turned off for shadow map vertex shaders?

The compiler should optimize all unused resources anyway, so it will change nothing.

> But the poly draw that's about to happen when the error is triggered is a shadow rendering shader that does not use the texture in question as a shader resource. Is there anything I can do to resolve the error?

Like I wrote:

> When you call CommitShaderResources(VERIFY), the engine verifies that all resources in the SRB are in SHADER_RESOURCE state. It can't know which resources will actually be needed by the PSO, so it verifies all. If some of the resources are not in SHADER_RESOURCE state, it will print a warning, which may be a false positive. In this case you need to disable verification and use the NONE mode.

**Comment by RobertMtx** [comment]
Okay, so its expected to receive error messages like this, even when the resource in question is not being used in the shader? That works for me. It certainly makes my job easier to just ignore it.

You may want to add a small note to the message to give a hint to developers that it may be safely ignored in some cases, or something along those lines. It shows up as a glaring red message in output/console with a label of "ERROR".  
![msg](https://github.com/DiligentGraphics/DiligentEngine/assets/43590381/2e9d82b1-8123-4a72-b045-828317e113fe)

Thanks for the help!

**Reply by TheMostDiligent** [reply]
You'd better use the `RESOURCE_STATE_TRANSITION_MODE_TRANSITION_NONE` mode to silence this error.
If there are layout mismatches, Direct3D debug layer will also print an error.
